ide/gototest/src/org/netbeans/api/gototest/TestOppositesLocator.java

```diff
@@ -0,0 +1,194 @@
+ * under the License.
+ */
+package org.netbeans.api.gototest;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.stream.Collectors;
+import org.netbeans.api.annotations.common.NonNull;
+import org.netbeans.spi.gototest.TestLocator;
+import org.netbeans.spi.gototest.TestLocator.LocationListener;
+import org.netbeans.spi.gototest.TestLocator.LocationResult;
+import org.openide.DialogDisplayer;
+import org.openide.NotifyDescriptor;
+import org.openide.filesystems.FileObject;
+import org.openide.util.Exceptions;
+import org.openide.util.Lookup;
+import org.openide.util.NbBundle;
+
+/**
+ * Find one or multiple test files for a source file,
+ * and one or multiple source files for a test file.
+ *
+ * @since 1.57
+ */
+public class TestOppositesLocator {
+
+    public static TestOppositesLocator getDefault() {
+        return new TestOppositesLocator();
+    }
+
+    /**
+     * Given the file and position in the file, if the:
+     * <ul>
+     *     <li> given file is a source file, find corresponding test file or test files, if exist.</li>
+     *     <li> given file is a test file, find corresponding source file or source files, if exist.</li>
+     * </ul>
+     *
+     * @param fo the file for which the opposites should be found
+     * @param caretOffset position in the file, or {@code -1} if unknown
+     * @return a result describing either an error, or a possibly empty list of locations found;
+     *         note one of {@code errorMessage} and {@code locations} is always {@code null},
+     *         and one always non-{@code null}.
+     */
+    @NbBundle.Messages("No_Test_Or_Tested_Class_Found=No Test or Tested class found")
+    public LocatorResult findOpposites(FileObject fo, int caretOffset) {
+        TestLocator.FileType currentFileType = getCurrentFileType(fo);
+        if(currentFileType == TestLocator.FileType.NEITHER) {
+            return new LocatorResult(Bundle.No_Test_Or_Tested_Class_Found(), null);
+        }
+        else {
+            return new LocatorResult(null, Collections.unmodifiableList(populateLocationResults(fo, caretOffset)));
+        }
+    }
+
+    private List<NamedLocation> populateLocationResults(FileObject fo, int caretOffset) {
+        Map<LocationResult, String> locationResults = new HashMap<LocationResult, String>();
+
+        Collection<? extends TestLocator> locators = Lookup.getDefault()
+                                                           .lookupAll(TestLocator.class)
+                                                           .stream()
+                                                           .filter(tl -> tl.appliesTo(fo))
+                                                           .collect(Collectors.toList());
+
+        CountDownLatch allDone = new CountDownLatch(locators.size());
```

Comment:
Will not work if one of the locators returns `appliesTo(fo) == false`; allDone still counts it.  Possible the `locator.appliesTo` test in the cycle is redundant, given they were already filtered.